src/middleware/CheckForDuplicateMiddleware.ts

```diff
@@ -0,0 +1,74 @@
+import { NextFunction, Request, Response } from 'express';
+import { UNAUTHORIZED } from 'http-status-codes';
+import { inject } from 'inversify';
+import { provide } from 'inversify-binding-decorators';
+import { BaseMiddleware } from 'inversify-express-utils';
+
+import { loggerInstance } from 'app/middleware/Logger';
+import { ApplicationData, APPLICATION_DATA_KEY } from 'app/models/ApplicationData';
+import { AppealsService } from 'app/modules/appeals-service/AppealsService';
+import { TOKEN_MISSING_ERROR } from 'app/utils/CommonErrors';
+import { sanitizeCompany } from 'app/utils/CompanyNumberSanitizer';
+import { getEnvOrThrow } from 'app/utils/EnvironmentUtils';
+
+const errorCustomTemplate: string = 'error-custom';
+const enquiryEmail: string = getEnvOrThrow('INQUIRY_EMAIL');
+
+const customErrorHeading = 'An Appeal has already been submitted for this penalty';
+const customErrorMessage = `If you think this is a mistake, email ${enquiryEmail}.`;
+
+@provide(CheckForDuplicateMiddleware)
+export class CheckForDuplicateMiddleware extends BaseMiddleware {
+
+    constructor(@inject(AppealsService) private readonly appealsService: AppealsService) {
+        super();
+    }
+
+    public async handler(request: Request, response: Response, next: NextFunction): Promise<void> {
+
+        if (!request.session) {
+            return next(new Error('Session is undefined'));
+        }
+
+        const signInInfo: ISignInInfo | undefined = request.session!.get<ISignInInfo>(SessionKey.SignInInfo);
+
+        const accessToken: string | undefined = signInInfo?.access_token?.access_token;
+
+        const refreshToken: string | undefined = signInInfo?.access_token?.refresh_token;
+
+        if (!accessToken) {
+            return next(TOKEN_MISSING_ERROR);
+        }
+
+        const applicationData: ApplicationData = request.session
+            .getExtraData(APPLICATION_DATA_KEY) || {} as ApplicationData;
+
+        const companyNumber: string = sanitizeCompany(applicationData.appeal.penaltyIdentifier.companyNumber);
+        const penaltyReference: string = applicationData.appeal.penaltyIdentifier.penaltyReference.toUpperCase();
+
+        loggerInstance()
+            .info(`CheckForDuplicateProcessor - Checking penalty ${penaltyReference} for duplicate appeals`);
+
+        try {
+            const isDuplicate: boolean = await this.appealsService
+                .hasExistingAppeal(companyNumber, penaltyReference, accessToken, refreshToken!);
+
+            console.log(isDuplicate);
+
+            if (isDuplicate) {
+                loggerInstance().error(`CheckForDuplicateProcessor - Duplicate appeal found for company ${companyNumber} and reference ${penaltyReference}`);
+                return response.status(UNAUTHORIZED).render(errorCustomTemplate, {
+                    heading: customErrorHeading,
+                    message: customErrorMessage
+                });
+            }
+        } catch (err) {
```

Comment:
I don't feel the try catch is necessary here - you're handling API call failures in your service class

Reply:
It's definitely necessary as if the api call fails it will throw an exception and ruin the express middleware call chain

Reply:
I think this is related to the next(err) that you mentioned previously

Reply:
thanks for pointing this out!